diff --git a/frameworks/av/media/libstagefright/ACodec.cpp b/frameworks/av/media/libstagefright/ACodec.cpp
--- a/frameworks/av/media/libstagefright/ACodec.cpp
+++ b/frameworks/av/media/libstagefright/ACodec.cpp
@@ -1377,7 +1377,13 @@
     status_t err = setComponentRole(encoder /* isEncoder */, mime);
 
     if (err != OK) {
-        return err;
+        if (!encoder
+                && !strcasecmp(mime, MEDIA_MIMETYPE_AUDIO_FLAC)
+                && !strcmp(mComponentName.c_str(), "OMX.google.flac.decoder")) {
+            err = OK;
+        } else {
+            return err;
+        }
     }
 
     int32_t bitRate = 0;
@@ -1854,11 +1860,11 @@
         } else {
             err = setupG711Codec(encoder, numChannels);
         }
-    } else if (encoder && !strcasecmp(mime, MEDIA_MIMETYPE_AUDIO_FLAC)) {
+    } else if (!strcasecmp(mime, MEDIA_MIMETYPE_AUDIO_FLAC)) {
         int32_t numChannels = 0, sampleRate = 0, compressionLevel = -1;
         if (!msg->findInt32("channel-count", &numChannels)
                     || !msg->findInt32("sample-rate", &sampleRate)) {
-            ALOGE("missing channel count or sample rate for FLAC encoder");
+            ALOGE("missing channel count or sample rate for FLAC codec");
             err = INVALID_OPERATION;
         } else {
             if (encoder) {
