--- a/build/core/main.mk	2026-09-15 00:15:56.350048979 +0200
+++ b/build/core/main.mk	2026-09-15 00:15:56.354049001 +0200
@@ -369,7 +369,9 @@

   ifeq ($(user_variant),userdebug)
     # Pick up some extra useful tools
+  ifneq ($(TARGET_PRODUCT),biscuit_minimal)
     tags_to_install += debug
+  endif
   else
     # Disable debugging in plain user builds.
     enable_target_debugging :=
@@ -864,9 +866,14 @@

 # All the droid stuff, in directories
 .PHONY: files
+ifeq ($(TARGET_BISCUIT_MINIMAL),true)
+files: $(modules_to_install) \
+        $(INSTALLED_ANDROID_INFO_TXT_TARGET)
+else
 files: prebuilt \
         $(modules_to_install) \
         $(INSTALLED_ANDROID_INFO_TXT_TARGET)
+endif

 # -------------------------------------------------------------------

--- a/build/core/Makefile	2026-09-15 00:15:56.354049001 +0200
+++ b/build/core/Makefile	2026-09-15 00:15:56.354049001 +0200
@@ -849,6 +849,9 @@

 SELINUX_FC := $(TARGET_ROOT_OUT)/file_contexts.bin
 INTERNAL_USERIMAGES_DEPS += $(SELINUX_FC)
+ifeq ($(TARGET_BISCUIT_MINIMAL),true)
+$(INSTALLED_RAMDISK_TARGET): $(SELINUX_FC)
+endif

 INTERNAL_USERIMAGES_DEPS += $(BLK_ALLOC_TO_BASE_FS)

@@ -866,6 +869,7 @@
 $(if $(BOARD_SYSTEMIMAGE_SQUASHFS_BLOCK_SIZE),$(hide) echo "system_squashfs_block_size=$(BOARD_SYSTEMIMAGE_SQUASHFS_BLOCK_SIZE)" >> $(1))
 $(if $(BOARD_SYSTEMIMAGE_SQUASHFS_DISABLE_4K_ALIGN),$(hide) echo "system_squashfs_disable_4k_align=$(BOARD_SYSTEMIMAGE_SQUASHFS_DISABLE_4K_ALIGN)" >> $(1))
 $(if $(PRODUCTS.$(INTERNAL_PRODUCT).PRODUCT_SYSTEM_BASE_FS_PATH),$(hide) echo "system_base_fs_file=$(PRODUCTS.$(INTERNAL_PRODUCT).PRODUCT_SYSTEM_BASE_FS_PATH)" >> $(1))
+$(if $(filter true,$(TARGET_BISCUIT_MINIMAL)),$(hide) echo "biscuit_minimal=true" >> $(1))
 $(if $(BOARD_USERDATAIMAGE_FILE_SYSTEM_TYPE),$(hide) echo "userdata_fs_type=$(BOARD_USERDATAIMAGE_FILE_SYSTEM_TYPE)" >> $(1))
 $(if $(BOARD_USERDATAIMAGE_PARTITION_SIZE),$(hide) echo "userdata_size=$(BOARD_USERDATAIMAGE_PARTITION_SIZE)" >> $(1))
 $(if $(filter true,$(TARGET_SKIP_USERDATAIMAGE)),$(hide) echo "skip_userdata_image=true" >> $(1))
@@ -1001,12 +1002,16 @@
 #   b) We build a single image that contains boot and recovery both
 #      (BOARD_USES_RECOVERY_AS_BOOT = true).

+ifeq ($(TARGET_BISCUIT_MINIMAL),true)
+RECOVERY_RESOURCE_ZIP :=
+else
 ifeq (,$(filter true, $(BOARD_USES_FULL_RECOVERY_IMAGE) $(BOARD_USES_RECOVERY_AS_BOOT) $(TARGET_NOT_USE_GZIP_RECOVERY_RAMDISK)))
 # Named '.dat' so we don't attempt to use imgdiff for patching it.
 RECOVERY_RESOURCE_ZIP := $(TARGET_OUT)/etc/recovery-resource.dat
 else
 RECOVERY_RESOURCE_ZIP :=
 endif
+endif

 ifeq ($(recovery_fstab),)
   $(info No recovery.fstab for TARGET_DEVICE $(TARGET_DEVICE))
@@ -1171,11 +1176,15 @@
 # image size check calculation.
 ifneq ($(INSTALLED_RECOVERYIMAGE_TARGET),)
 intermediates := $(call intermediates-dir-for,PACKAGING,recovery_patch)
+ifeq ($(TARGET_BISCUIT_MINIMAL),true)
+RECOVERY_FROM_BOOT_PATCH :=
+else
 ifndef BOARD_CANT_BUILD_RECOVERY_FROM_BOOT_PATCH
 RECOVERY_FROM_BOOT_PATCH := $(intermediates)/recovery_from_boot.p
 else
 RECOVERY_FROM_BOOT_PATCH :=
 endif
+endif
 $(RECOVERY_FROM_BOOT_PATCH): $(INSTALLED_RECOVERYIMAGE_TARGET) \
                              $(INSTALLED_BOOTIMAGE_TARGET) \
 			     $(HOST_OUT_EXECUTABLES)/imgdiff \
@@ -1237,12 +1246,18 @@
     $(filter-out $(foreach p,$(overridden_packages),$(p) %/$(p).apk), \
         $(ALL_PDK_FUSION_FILES))

+ifeq ($(TARGET_BISCUIT_MINIMAL),true)
+INTERNAL_SYSTEMIMAGE_FILES := $(filter $(TARGET_OUT)/%, \
+    $(ALL_DEFAULT_INSTALLED_MODULES) \
+    $(PDK_FUSION_SYSIMG_FILES))
+else
 INTERNAL_SYSTEMIMAGE_FILES := $(filter $(TARGET_OUT)/%, \
     $(ALL_PREBUILT) \
     $(ALL_GENERATED_SOURCES) \
     $(ALL_DEFAULT_INSTALLED_MODULES) \
     $(PDK_FUSION_SYSIMG_FILES) \
     $(RECOVERY_RESOURCE_ZIP))
+endif


 FULL_SYSTEMIMAGE_DEPS := $(INTERNAL_SYSTEMIMAGE_FILES) $(INTERNAL_USERIMAGES_DEPS)
@@ -1296,10 +1311,49 @@
 endef
 endif

+ifeq ($(TARGET_BISCUIT_MINIMAL),true)
+define clean-biscuit-minimal-target-out
+  $(hide) rm -rf \
+    $(TARGET_OUT)/etc/install-recovery.sh \
+    $(TARGET_OUT)/etc/recovery-resource.dat \
+    $(TARGET_OUT)/recovery-from-boot.p \
+    $(TARGET_OUT)/usr/icu \
+    $(TARGET_OUT)/usr/share/vim \
+    $(TARGET_OUT)/bin/app_process32 \
+    $(TARGET_OUT)/bin/asan/app_process32 \
+    $(TARGET_OUT)/bin/bootanimation \
+    $(TARGET_OUT)/bin/screencap \
+    $(TARGET_OUT)/bin/minikin_skia_example \
+    $(TARGET_OUT)/bin/bcc \
+    $(TARGET_OUT)/bin/ld.mc \
+    $(TARGET_OUT)/bin/llvm-symbolizer \
+    $(TARGET_OUT)/lib/libLLVM.so \
+    $(TARGET_OUT)/lib/libRSDriver.so \
+    $(TARGET_OUT)/lib/libRS_internal.so \
+    $(TARGET_OUT)/lib/libandroid_runtime.so \
+    $(TARGET_OUT)/lib/libandroidfw.so \
+    $(TARGET_OUT)/lib/libbcc.so \
+    $(TARGET_OUT)/lib/libbcinfo.so \
+    $(TARGET_OUT)/lib/libdng_sdk.so \
+    $(TARGET_OUT)/lib/libft2.so \
+    $(TARGET_OUT)/lib/libgui.so \
+    $(TARGET_OUT)/lib/libhwui.so \
+    $(TARGET_OUT)/lib/libicui18n.so \
+    $(TARGET_OUT)/lib/libicuuc.so \
+    $(TARGET_OUT)/lib/libmedia.so \
+    $(TARGET_OUT)/lib/libpdfium.so \
+    $(TARGET_OUT)/lib/libskia.so
+endef
+else
+define clean-biscuit-minimal-target-out
+endef
+endif
+
 # $(1): output file
 define build-systemimage-target
   @echo "Target system fs image: $(1)"
   $(call create-system-vendor-symlink)
+  $(call clean-biscuit-minimal-target-out)
   @mkdir -p $(dir $(1)) $(systemimage_intermediates) && rm -rf $(systemimage_intermediates)/system_image_info.txt
   $(call generate-userimage-prop-dictionary, $(systemimage_intermediates)/system_image_info.txt, \
       skip_fsck=true)
--- a/build/tools/releasetools/common.py	2026-09-15 00:15:56.358049025 +0200
+++ b/build/tools/releasetools/common.py	2026-09-15 00:15:56.358049025 +0200
@@ -1764,6 +1764,9 @@
   if info_dict is None:
     info_dict = OPTIONS.info_dict

+  if info_dict.get("biscuit_minimal", None) == "true":
+    return
+
   full_recovery_image = info_dict.get("full_recovery_image", None) == "true"
   system_root_image = info_dict.get("system_root_image", None) == "true"
   use_bsdiff = info_dict.get("no_gzip_recovery_ramdisk", None) == "true"
