--- a/frameworks/base/data/keyboards/Generic.kl
+++ b/frameworks/base/data/keyboards/Generic.kl
@@ -132,7 +132,7 @@
 key 110   INSERT
 key 111   FORWARD_DEL
 # key 112 "KEY_MACRO"
-key 113   VOLUME_MUTE
+key 113   MUTE
 key 114   VOLUME_DOWN
 key 115   VOLUME_UP
 key 116   POWER
@@ -157,7 +157,7 @@
 key 135   PASTE
 # key 136 "KEY_FIND"
 key 137   CUT
-# key 138 "KEY_HELP"
+key 138   HELP
 key 139   MENU
 key 140   CALCULATOR
 # key 141 "KEY_SETUP"
--- a/frameworks/base/services/core/java/com/android/server/policy/PhoneWindowManager.java
+++ b/frameworks/base/services/core/java/com/android/server/policy/PhoneWindowManager.java
@@ -6850,11 +6850,20 @@
                 break;
             }
 
+            case KeyEvent.KEYCODE_MUTE: {
+                result &= ~ACTION_PASS_TO_USER;
+                if (down && event.getRepeatCount() == 0) {
+                    AudioManager audioManager =
+                            (AudioManager) mContext.getSystemService(Context.AUDIO_SERVICE);
+                    audioManager.setMicrophoneMute(!audioManager.isMicrophoneMute());
+                }
+                break;
+            }
+
             case KeyEvent.KEYCODE_MEDIA_PLAY:
             case KeyEvent.KEYCODE_MEDIA_PAUSE:
             case KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE:
             case KeyEvent.KEYCODE_HEADSETHOOK:
-            case KeyEvent.KEYCODE_MUTE:
             case KeyEvent.KEYCODE_MEDIA_STOP:
             case KeyEvent.KEYCODE_MEDIA_NEXT:
             case KeyEvent.KEYCODE_MEDIA_PREVIOUS:
--- a/frameworks/base/services/core/java/com/android/server/audio/AudioService.java
+++ b/frameworks/base/services/core/java/com/android/server/audio/AudioService.java
@@ -1965,6 +1965,14 @@
         sendStickyBroadcastToAll(intent);
     }
 
+    private void broadcastMicrophoneMuteStatus(boolean muted) {
+        Intent intent = new Intent("com.amazon.biscuit.service.MICROPHONE_MUTE_CHANGED");
+        intent.putExtra("com.amazon.biscuit.service.EXTRA_MICROPHONE_MUTED", muted);
+        intent.addFlags(Intent.FLAG_RECEIVER_REGISTERED_ONLY_BEFORE_BOOT
+                | Intent.FLAG_RECEIVER_REPLACE_PENDING);
+        sendStickyBroadcastToAll(intent);
+    }
+
     /**
      * Sets the stream state's index, and posts a message to set system volume.
      * This will not call out to the UI. Assumes a valid stream type.
@@ -2246,9 +2254,12 @@
         }
         // If mute is for current user actually mute, else just persist the setting
         // which will be loaded on user switch.
+        boolean oldMuted = AudioSystem.isMicrophoneMuted();
         if (getCurrentUserId() == userId) {
             AudioSystem.muteMicrophone(on);
         }
+        boolean muted = AudioSystem.isMicrophoneMuted();
+        if (oldMuted != muted) broadcastMicrophoneMuteStatus(muted);
         // Post a persist microphone msg.
     }
 
