--- a/device/amazon/mt8163-common/rootdir/etc/init.mt8163.rc
+++ b/device/amazon/mt8163-common/rootdir/etc/init.mt8163.rc
@@ -505,0 +506,23 @@
+service wmt_loader /vendor/bin/wmt_loader
+    class core
+    user root
+    group root system
+    oneshot
+    disabled
+
+service wmt_launcher /vendor/bin/wmt_launcher -p /vendor/firmware/
+    class core
+    user system
+    group system
+    disabled
+
+on property:ro.product.device=biscuit
+    chmod 0660 /dev/stpwmt
+    chmod 0660 /dev/wmtWifi
+    chmod 0660 /dev/stpbt
+    chown system system /dev/stpwmt
+    chown system system /dev/wmtWifi
+    chown bluetooth bluetooth /dev/stpbt
+    start wmt_loader
+    start wmt_launcher
+
