--- a/system/bt/bta/dm/bta_dm_act.c
+++ b/system/bt/bta/dm/bta_dm_act.c
@@ -2931,6 +2931,7 @@ static UINT8 bta_dm_authentication_complete_cback(BD_ADDR bd_addr, DEV_CLASS dev
 ** Returns          void
 **
 *******************************************************************************/
+#if (BTM_LOCAL_IO_CAPS != BTM_IO_CAP_NONE)
 static UINT8 bta_dm_sp_cback (tBTM_SP_EVT event, tBTM_SP_EVT_DATA *p_data)
 {
     tBTM_STATUS status = BTM_CMD_STARTED;
@@ -3090,6 +3091,7 @@ static UINT8 bta_dm_sp_cback (tBTM_SP_EVT event, tBTM_SP_EVT_DATA *p_data)
     APPL_TRACE_EVENT("dm status: %d", status);
     return status;
 }
+#endif
 
 /*******************************************************************************
 **
